### DISCUSSION TITLE
Show more information about exceptions failing tests

### DISCUSSION THREAD

**Discussion by dave-yotta** [nunit/nunit#4858] [How to]
Have had a look for a bit now and am going to just wrap some cases with `try { .. } catch(Exception e) { logger.Log(SerializeException(d)); throw }`.  I've got more data on my subclass of `Exception` that will tell me what the issue is, but <ErrorInfo> only contains the message and stack trace of the top level exception.

Be helpful if there was a way to have `string AddResultInfo(Exception testException)` (but better thought out :D) so I can enable this information globally.

Am I missing something?

**Comment by manfred-brands** [comment]
I didn't know this before, but NUnit does show the contents of the [Exception.Data[(https://learn.microsoft.com/en-us/dotnet/api/system.exception.data?view=net-8.0) dictionary.

It does this for all inner exceptions.

The code segment below gives this output:
```
Message: 
  System.AggregateException : Something went wrong
    ----> System.ArgumentException : First argument is wrong
  Parameter name: firstArg
    ----> System.InvalidOperationException : You can't do that
    ----> NUnit.Framework.Tests.Assertions.AssertThrowsTests+MyException : my message
  Data:
    42: TheAnswer
    0: Zero
    ----> NUnit.Framework.Tests.Assertions.AssertThrowsTests+MyException : my other message
  Data:
    pi: 3.14159
    e: 2.71828
```

Code:
```csharp
public class TestExceptionData
{
    [Test]
    public void ShowsAllData()
    {
        Dictionary<int, string> data = new Dictionary<int, string>
        {
            { 42, "TheAnswer" },
            { 0, "Zero" }
        };

        Dictionary<string, double> otherData = new Dictionary<string, double>
        {
            { "pi", 3.14159 },
            { "e", 2.71828 }
        };

        throw new AggregateException("Something went wrong",
            new ArgumentException("First argument is wrong", "firstArg"),
            new InvalidOperationException("You can't do that"),
            new MyException("my message", data),
            new MyException("my other message", otherData));
    }

    private sealed class MyException : Exception
    {
        public MyException(string message, IDictionary data)
            : base(message)
        {
            Data = data;
        }

        public override IDictionary Data { get; }
    }
```

**Comment by dave-yotta** [comment]
That's interesting! We're not using that though, we have some other properties on the exception class and don't really want to set `System.Exception.Data` in production :)

**Comment by manfred-brands** [comment]
`System.Exception.Data` is a virtual get only property.
Your exception class can implement it to return your property names and values.

**Reply by dave-yotta** [reply]
I know I know, but this would be something solely for test purposes that's being put into production code so I don't like the solution :)

I wouldn't want to filter out duplicate data for example going to monitoring like sentry.

Thanks for the suggestion though!

**Comment by stevenaw** [comment]
You are not the first or even second person to find this need. :)
This has been requested as a feature within the framework itself in the past, most recently in https://github.com/nunit/nunit/issues/4128

One option suggested there is to use a tool like `Verify`. Could that work until the team decides on a direction for anything within NUnit itself?

**Reply by dave-yotta** [reply]
Managed to get by this time; usually if there's not enough info on the test result we'll debug it...but this one was only happening in CI for some reason.

While it's not a http call we're testing in this instance, your comment does make me wonder if `throw new xxxException` is a code smell like you have with `new` itself and factories.  If we had `IExceptionThrowingFactory` then these would have been able to add more data to the test logs by decorating it during tests.
